### DISCUSSION TITLE
Waypoints list

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#114] [Q&A] [answered]
The new/larger icons UI is very good and does make some actions easier.

I've mentioned this one previously, but the Waypoints list isn't great in the landscape orientation. The list isn't big enough to scroll and I think there's wasted screen real-estate.  Perhaps buttons on the left/right so the list can use the full "height" of the window when in landscape?  I'm guessing at least 3 list items visible would be ideal.

Maybe remove the Shaping/Via-Point button and have that as the list item's icon which you can touch to toggle?

Single-line distance/time as there's lots of room along the single line :)

What I see on my Samsung S20:
![waypoints-landscape](https://user-images.githubusercontent.com/125696109/229625680-2dfad326-8b36-4ca6-b3e2-15bfce4de146.jpg)

I should point out that I'm using a new phone mount (sprung-loaded holder) which sits behind my screen BMW Navigator style, via an AMPS plate - so I can't rotate it without taking it out of the cradle to hold it :)  My thinking will most-likely be sideways from now on!

_Originally posted by @MotoUKRider in https://github.com/devemux86/cruiser/discussions/82#discussioncomment-5514491_

---

For the new implementation see here:
- https://github.com/devemux86/cruiser/discussions/114#discussioncomment-5559274

_Edited by @devemux86_

**Comment by devemux86** [comment]
is drag and drop possible in this menu to be able to move points?

_Originally posted by @DolomitesTom in https://github.com/devemux86/cruiser/discussions/82#discussioncomment-5516928_

**Comment by devemux86** [comment]
@MotoUKRider thanks for the suggestion!

You are right, the waypoint list can and should be improved and definitely needs better organization in the landscape. :+1: 

I have put it in my implementation list to review its design.

**Reply by MotoUKRider** [reply]
Thank you :)

I also had a thought about buttons-vs-space: How about a long-press on the item to delete it, with a "large" yes/no confirmation box?  That adds a safety-net aspect to big-fingered navigation and frees-up some space.  Or, have the "delete" button at the far-right end of the list row?

Or... all those buttons placed over on the right-hand-side in a 2x3 formation so the list is the vertical size of the popup dialog?

Hopefully the difference between portrait+lanscape in coding terms isn't too much work!

**Reply by devemux86** [reply]
The long press menu is certainly an interesting idea. :+1: 

It can free the toolbar from some buttons and a menu can host more new actions in the future.

**Comment by devemux86** [comment]
> is drag and drop possible in this menu to be able to move points?

@DolomitesTom thanks for the idea!

Drag and drop is already in my implementation list, you can expect it in the future!

**Comment by devemux86** [comment]
> Single-line distance/time as there's lots of room along the single line :)

This is indeed a quick first improvement in landscape orientation:

![Screenshot_1680613249](https://user-images.githubusercontent.com/3484020/229799986-5c8bc454-26f4-4f5b-8798-6b7a063d855f.jpg)

**Reply by MotoUKRider** [reply]
Yes, that looks better. Makes it easier to scroll and see what is above/below when using the up/down "move" buttons.

**Reply by devemux86** [reply]
> Single-line distance/time as there's lots of room along the single line :)

Available in [Cruiser 3.0.17](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by DolomitesTom** [reply]
Still not that convenient with big font settings
![Screenshot_20230406-161545_Cruiser](https://user-images.githubusercontent.com/121371629/230405110-728737d1-01e2-490c-80c6-8b72cfc5fb17.jpg)
Maybe better with ok and cancel button in the icon line and without a headline.

**Reply by devemux86** [reply]
Indeed, we could use more efficiently the available space.
The one line of waypoint info is a quick first improvement.

In the long run we could use a new design without toolbar,
but with drag and drop and (long press) menu on waypoints.

**Reply by DolomitesTom** [reply]
Yes

**Reply by WolfgangZ42** [reply]
Not to bad in the first run:
 
<img width="400" src="https://user-images.githubusercontent.com/121806238/230424512-73e09f90-002f-432f-999f-6065f8dac4cc.png"/>

**Reply by MotoUKRider** [reply]
I agree - a nice "quick" fix, looks much better.

CoPilot has drag'n'drop on its version of this screen, but is very hit'n'miss with items not always moving and the dragged outline of the item sometimes "ghosting" with a re-draw issue (in coder's speak!).  The up/down buttons work great, and I'm sure the dragging method would also be welcome by others.

I favour the "stabby" method of touch, long-press, toggle up/down etc.  I'm sure drag'n'drop would be popular with the portrait users :)

**Comment by devemux86** [comment]
Drag and drop is often one of the most complicated things to get right in all programming languages.
I had already implemented it on desktop Cruiser, I can enable it + right click menu in the next update.

And I will see how I can implement it on Android Cruiser as well. :slightly_smiling_face:

**Comment by devemux86** [comment]
New waypoints list coming in a next Cruiser (see video):
- Drag and drop: reorder waypoints
(drag handler is the icon)
- Swipe left / right: delete waypoints
- Tap: change via <-> shaping point
- Long press: menu
  - Segment profile
  - Edit
  - Show on map
  - Select (opens actions menu)
  - Change all (via <-> shaping points)

[device-2023-04-08-113924.webm](https://user-images.githubusercontent.com/3484020/230716713-5688d110-9e27-4215-9e1a-10fb208ff6d3.webm)

**Reply by devemux86** [reply]
I updated [above](https://github.com/devemux86/cruiser/discussions/114#discussioncomment-5559274) all the actions in the new waypoints list.

Please check them and see the video.
Ιs there anything else needed?

**Reply by MotoUKRider** [reply]
That all seems quite well-thought out!

**Reply by devemux86** [reply]
We can also use the "long press" to initiate the drag instead of the icon,
but we need the "long press" to activate the menu with the extra actions.

**Reply by 0709wiwiwi** [reply]
It may work fine, but I already had to think how to activate the more menu.  I suppose a newbie will not discover this  very easily.  
What is described [HERE](https://github.com/devemux86/cruiser/discussions/114#discussioncomment-5559274) will soon be lost sight of too quickly without a hint into the Cruiser wiki page. (_The less user manual needed the better_)

**Reply by devemux86** [reply]
I thought about adding drag indicators (there are specific icons), but it would eat up horizontal space on each line?

![drag_handle_black_24dp](https://user-images.githubusercontent.com/3484020/236865109-42852b5e-7b26-4522-8dd1-52fb42aea14f.png) ![drag_indicator_black_24dp](https://user-images.githubusercontent.com/3484020/236865266-32971cef-2427-441d-9d53-5ad039e109a1.png)

Like the Languages screen in Android settings:

<img src="https://user-images.githubusercontent.com/3484020/236866351-d428edc3-925d-4e84-a326-3685c55583f1.jpg" width="33%"/>

**Reply by 0709wiwiwi** [reply]
Shifting through the Icons still seems discoverable to me.  Fine ok.
And previous version users should know that you can bulk convert ALL Shaping <=> Via points. 
How would a newbie discover this hidden gem ? Long press waypoint > change > = Action menu.

**Reply by devemux86** [reply]
Tap on a waypoint changes its type: via <=> shaping.

**Reply by 0709wiwiwi** [reply]
This indeed works perfectly Emux. Nothing to comment on. I do suspect that possibly not all are familiar with the long push usage in Cruiser as more menu choice. Without overloading the Cruiser wiki it could be reported that short or long push offers menu variation. (A bit like Windows mouse right clicks)

**Comment by devemux86** [comment, marked as answer]
New waypoints list with drag & drop is available in [Cruiser 3.0.18](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

For instructions please see above:
- https://github.com/devemux86/cruiser/discussions/114#discussioncomment-5559274

Please let me know how you see it!

**Reply by DolomitesTom** [reply]
Delightful, works perfectly

**Reply by MotoUKRider** [reply]
Tried this a few times - works nice.  Thanks.
